File: Sources/Sentry/SentrySpanOperation.m
#import "SentrySpanOperation.h"

NSString *const SentrySpanOperationAppLifecycle = @"app.lifecycle";

NSString *const SentrySpanOperationCoredataFetchOperation = @"db.sql.query";
NSString *const SentrySpanOperationCoredataSaveOperation = @"db.sql.transaction";

NSString *const SentrySpanOperationFileRead = @"file.read";
NSString *const SentrySpanOperationFileWrite = @"file.write";
NSString *const SentrySpanOperationFileCopy = @"file.copy";
NSString *const SentrySpanOperationFileRename = @"file.rename";
NSString *const SentrySpanOperationFileDelete = @"file.delete";

NSString *const SentrySpanOperationNetworkRequestOperation = @"http.client";

NSString *const SentrySpanOperationUiAction = @"ui.action";
NSString *const SentrySpanOperationUiActionClick = @"ui.action.click";

NSString *const SentrySpanOperationUiLoad = @"ui.load";
NSString *const SentrySpanOperationUiLoadInitialDisplay = @"ui.load.initial_display";
NSString *const SentrySpanOperationUiLoadFullDisplay = @"ui.load.full_display";

@implementation SentrySpanOperation

+ (NSString *)appLifecycle
{
    return SentrySpanOperationAppLifecycle;
}

+ (NSString *)coredataFetchOperation
{
    return SentrySpanOperationCoredataFetchOperation;
}

+ (NSString *)coredataSaveOperation
{
    return SentrySpanOperationCoredataSaveOperation;
}

+ (NSString *)fileRead
{
    return SentrySpanOperationFileRead;
}

+ (NSString *)fileWrite
{
    return SentrySpanOperationFileWrite;
}

+ (NSString *)fileCopy
{
    return SentrySpanOperationFileCopy;
}

+ (NSString *)fileRename
{
    return SentrySpanOperationFileRename;
}

+ (NSString *)fileDelete
{
    return SentrySpanOperationFileDelete;
}

+ (NSString *)networkRequestOperation
{
    return SentrySpanOperationNetworkRequestOperation;
}

+ (NSString *)uiAction
{
    return SentrySpanOperationUiAction;
}

+ (NSString *)uiActionClick
{
    return SentrySpanOperationUiActionClick;
}

+ (NSString *)uiLoad
{
    return SentrySpanOperationUiLoad;
}

+ (NSString *)uiLoadInitialDisplay
{
    return SentrySpanOperationUiLoadInitialDisplay;
}

+ (NSString *)uiLoadFullDisplay
{
    return SentrySpanOperationUiLoadFullDisplay;
}
@end

File: Sources/Swift/Transactions/SentrySpanOperation.swift
import Foundation

/**
 * Span operations are short string identifiers that categorize the type of operation a span is measuring.
 *
 * They follow a hierarchical dot notation format (e.g. `ui.load.initial_display`) to group related operations.
 * These identifiers help organize and analyze performance data across different types of operations.
 *
 *  - Remark: As  Swift `enum` are not available in Objective-C, it uses a `class` with static properties, marked with `@objc` and `@objcMembers` instead.
 *          To reduce casting between `String` to `NSString` when using from Objective-C, it uses `NSString` instead of `String`.
 *          Eventually this should be replaced with a Swift `enum` when Objective-C compatibility is not needed anymore.
 *  - Note: See [Sentry SDK development documentation](https://develop.sentry.dev/sdk/telemetry/traces/span-operations/) for more information.
 */
@objcMembers @objc(SentrySpanOperation)
class SentrySpanOperation: NSObject {
    static let appLifecycle: NSString = "app.lifecycle"

    static let coredataFetchOperation: NSString = "db.sql.query"
    static let coredataSaveOperation: NSString = "db.sql.transaction"

    static let fileRead: NSString = "file.read"
    static let fileWrite: NSString = "file.write"
    static let fileCopy: NSString = "file.copy"
    static let fileRename: NSString = "file.rename"
    static let fileDelete: NSString = "file.delete"

    static let networkRequestOperation: NSString = "http.client"

    static let uiAction: NSString = "ui.action"
    static let uiActionClick: NSString = "ui.action.click"

    static let uiLoad: NSString = "ui.load"
    static let uiLoadInitialDisplay: NSString = "ui.load.initial_display"
    static let uiLoadFullDisplay: NSString = "ui.load.full_display"
}
